Securing your Buildkite Agent and environment variables (#2761)
buildkite/docs: pages/agent/v3/securing.md

# Securing your Buildkite Agent

> In cases where a Buildkite Agent is being deployed into a sensitive environment there are a few default settings which may be adjusted and techniques which may be used.

Reviewer: ```suggestion
In cases where a Buildkite Agent is being deployed into a sensitive environment, there are a few default settings which may be adjusted and techniques that may be used.
```

* Command line flag: `--no-command-eval`
* Configuration setting: `no-command-eval=true`

> 🚧 Custom hooks and environment variables
> If you have a custom `command` hook, using `no-command-eval` will have no effect on your command execution. See [Allowing a list of plugins](#allowing-a-list-of-plugins) and [Custom bootstrap scripts](#customizing-the-bootstrap) for examples of how to completely lock down your agent from arbitrary code execution.
>

> > Using `no-command-eval` only prevents command evaluation by the agent itself. Other programs that run during the job could be influenced into executing arbitrary commands via environment variables (for example, `BASH_ENV` or `GIT_SSH_COMMAND`). See [Strict checks using a pre-bootstrap hook](#strict-checks-using-a-pre-bootstrap-hook) and [`enable-environment-variable-allowlist`](/docs/agent/v3/cli-start#enable-environment-variable-allowlist) for possible approaches to filtering environment variables.

Reviewer: Following on from your last comment then ...

```suggestion
> Using `no-command-eval` only prevents command evaluation by the agent itself. Other programs and tools that run during the job (for example, build or test tools) could be influenced into executing arbitrary commands via environment variables (for example, `BASH_ENV` or `GIT_SSH_COMMAND`). See [Strict checks using a pre-bootstrap hook](#strict-checks-using-a-pre-bootstrap-hook) and [`enable-environment-variable-allowlist`](/docs/agent/v3/cli-start#enable-environment-variable-allowlist) for possible approaches to filtering environment variables.
```

> >🚧 Custom hooks
> > If you have a custom <code>command</code> hook, using <code>no-command-eval</code> will have no effect on your command execution. See <a href="#allowing-a-list-of-plugins">Allowing a list of plugins</a> and <a href="#customizing-the-bootstrap">Custom bootstrap scripts</a> for examples of how to completely lock down your agent from arbitrary code execution.

Reviewer: Switch to using Markdown syntax, which [can be used in callouts](https://github.com/buildkite/docs/blob/main/styleguides/markdown-syntax-style.md#callouts) as doing so makes the source content more readable.

```suggestion
> 🚧 Custom hooks
> If you have a custom `command` hook, using `no-command-eval` will have no effect on your command execution. See [Allowing a list of plugins](#allowing-a-list-of-plugins) and [Custom bootstrap scripts](#customizing-the-bootstrap) for examples of how to completely lock down your agent from arbitrary code execution.
```

> <!-- comment to ensure we create two separate callout blocks -->
> 
> >🚧 Environment variables
> > Using <code>no-command-eval</code> only prevents command evaluation by the agent itself. Other programs that run during the job could be influenced into executing arbitrary commands via environment variables (for example, `BASH_ENV` or `GIT_SSH_COMMAND`). See <a href="#strict-checks-using-a-pre-bootstrap-hook">Strict checks using a pre-bootstrap hook</a> and <a href="/docs/agent/v3/cli-start#enable-environment-variable-allowlist"><code>enable-environment-variable-allowlist</code></a> for possible approaches to filtering environment variables.

Reviewer: Following on from above ...

```suggestion
<!-- Markdown comment divider create two separate callout blocks -->

> 🚧 Environment variables
> Using `no-command-eval` only prevents command evaluation by the agent itself. Other programs that run during the job could be influenced into executing arbitrary commands via environment variables (for example, `BASH_ENV` or `GIT_SSH_COMMAND`). See [Strict checks using a pre-bootstrap hook](#strict-checks-using-a-pre-bootstrap-hook) and [`enable-environment-variable-allowlist`](/docs/agent/v3/cli-start#enable-environment-variable-allowlist) for possible approaches to filtering environment variables.
```

When you say "Other programs", what exactly is meant by 'programs'? What might be an example of such a program?

> >🚧 Custom hooks
> > If you have a custom <code>command</code> hook, using <code>no-command-eval</code> will have no effect on your command execution. See <a href="#allowing-a-list-of-plugins">Allowing a list of plugins</a> and <a href="#customizing-the-bootstrap">Custom bootstrap scripts</a> for examples of how to completely lock down your agent from arbitrary code execution.
> 
> <!-- comment to ensure we create two separate callout blocks -->
> 
> >🚧 Environment variables
> > Using <code>no-command-eval</code> only prevents command evaluation by the agent itself. Other programs that run during the job could be influenced into executing arbitrary commands via environment variables (for example, `BASH_ENV` or `GIT_SSH_COMMAND`). See <a href="#strict-checks-using-a-pre-bootstrap-hook">Strict checks using a pre-bootstrap hook</a> and <a href="/docs/agent/v3/cli-start#enable-environment-variable-allowlist"><code>enable-environment-variable-allowlist</code></a> for possible approaches to filtering environment variables.

Reviewer: Rather than writing two consecutive callouts, which can look a tad naff in the docs, perhaps combine these into one with separate paragraphs. Also, switch to using Markdown syntax, which [can be used in callouts](https://github.com/buildkite/docs/blob/main/styleguides/markdown-syntax-style.md#callouts) as doing so makes the source content more readable ...

 ```suggestion
> 🚧 Custom hooks and environment variables
> If you have a custom `command` hook, using `no-command-eval` will have no effect on your command execution. See [Allowing a list of plugins](#allowing-a-list-of-plugins) and [Custom bootstrap scripts](#customizing-the-bootstrap) for examples of how to completely lock down your agent from arbitrary code execution.
>
> Using `no-command-eval` only prevents command evaluation by the agent itself. Other programs that run during the job could be influenced into executing arbitrary commands via environment variables (for example, `BASH_ENV` or `GIT_SSH_COMMAND`). See [Strict checks using a pre-bootstrap hook](#strict-checks-using-a-pre-bootstrap-hook) and [`enable-environment-variable-allowlist`](/docs/agent/v3/cli-start#enable-environment-variable-allowlist) for possible approaches to filtering environment variables.
```

Last query, when you say "Other programs", what exactly is meant by 'programs'? What might be an example of such a program?

Author: > Last query, when you say "Other programs", what exactly is meant by 'programs'? What might be an example of such a program?

The example environment variables relate to Bash and Git, but I want to communicate the possibility that any build/test/other tool that the customer is using in a job might be exploitable in this way.
